### ISSUE TITLE
External plugin system rework

### ISSUE THREAD

**Issue by grejppi** [LMMS/lmms#2616]
**TL;DR** -- plans to redesign how LMMS does external plugins so that these can become a reality:
- LV2 support: #562
- LinuxVST support: #416

_The old, incoherent rant written while sleep deprived has been moved to https://gist.github.com/grejppi/0f1063e8a08137e23e8cefeab035f9fa for archival purposes_

---

I have taken the liberty of starting some preliminary work on what could become a new unified way of loading external plugins as instruments and effects. (The code is [here](https://github.com/grejppi/lmms/tree/plugin-rework))
- `ExternalEffect` class replaces the existing `LadspaEffect` & `VstEffect` classes.
- `ExternalInstrument` class is introduced, which is a new LMMS instrument.
- `ExternalPlugin` class is introduced, which is an abstraction of different plugin formats. It provides virtual functions to discover and run plugins.

**Comment by x42** [comment]
What's the issue with LV2 that lead to despair?

It's really straight forward, Most if not all of it is abstracted by liblilv  and the GUI part by libsuil. It works cross platform on the four+ major platforms out there.

Yet another plugin format? anything innovative?

**Comment by grejppi** [comment]
@x42 the despair was not caused by LV2, but the fact that LMMS reimplements the same things for every plugin format.

**Comment by paper-pauper** [comment]
LV2 is standard, and supported by plenty of existing programs. I suggest not only that LMMS implements LV2 support, but also that it rewrites its own instruments (FreeBoy, Kicker, LB302, etc.) as LV2 plugins, so that they can be used outside of LMMS.

Someone has already done some ground work for the second point: [here](https://github.com/pgiblock/lmms-lv2).

I think this issue is very important, so much that you should probably start a crowdfunding campaign for it. This would make the difference between a hobby program and a professional one.

**Comment by tresf** [comment]
> I think this issue is very important, so much that you should probably start a crowdfunding campaign for it.

Please do.  This isn't a Sunday afternoon task and it's easier to talk about it than it is to do it.  If you want to step in and coordinate, 1. Prove you can. 2. Do it.  Otherwise you're just saying `+1`.

**Comment by paper-pauper** [comment]
How am I, who's never contributed to LMMS or been involved in its development, to be trusted with raising funds for it?

If a trustworthy LMMS developer started a crowdfunding campaign I would gladly donate to it and promote it, but please don't expect people who just express an opinion to take responsibilities they don't have the authority to claim for themselves.

**Comment by tresf** [comment]
@paper-pauper, to keep the thread on topic, I'll ask you bring you "opinions" to the development list or a separate bug report.

**Comment by paper-pauper** [comment]
@tresf You went off thread by the way, by using an ad hominem. Consider my former post as a `+1` if you wish and it's still a valid comment, when both yours are useless and really bring nothing to the discussion.

**Comment by tresf** [comment]
@paper-pauper I've started a discussion here:  https://github.com/LMMS/lmms/issues/2745. I've closed it out so that it won't show as open on our tracker, but comments are enabled.

**Comment by karmux** [comment]
I see you've deleted branch with preliminary changes. Are you still working on this?
